File: packages/guards/src/data-poisoning.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface DataPoisoningOptions { action: 'block' | 'warn'; }

const POISONING_PATTERNS: RegExp[] = [
  /\b(?:from\s+now\s+on|going\s+forward|remember\s+that|always\s+(?:say|respond|answer))\b/gi,
  /\b(?:whenever\s+(?:someone|anyone|a\s+user)\s+asks?|if\s+(?:asked|prompted)\s+about)\b/gi,
  /\b(?:your\s+new\s+(?:instruction|rule|policy|behavior|personality)\s+is)\b/gi,
  /\b(?:update|change|modify|override)\s+your\s+(?:instructions|rules|system\s+prompt|behavior)\b/gi,
  /\b(?:insert|add|append|inject)\s+(?:this|the\s+following)\s+(?:into|to)\s+your\s+(?:memory|knowledge|training|context)\b/gi,
  /\b(?:learn|memorize|internalize)\s+(?:this|that|the\s+following)\b/gi,
  /\b(?:you\s+(?:were|are)\s+(?:trained|programmed|instructed)\s+to)\b/gi,
];

export function dataPoisoning(options: DataPoisoningOptions): Guard {
  return {
    name: 'data-poisoning', version: '0.1.0',
    description: 'Detect attempts to poison or alter LLM behavior/knowledge',
    category: 'security', supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];
      for (const p of POISONING_PATTERNS) {
        const re = new RegExp(p.source, p.flags);
        const m = re.exec(text);
        if (m) matched.push(m[0]);
      }
      const triggered = matched.length > 0;
      return {
        guardName: 'data-poisoning', passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered ? `Data poisoning attempt: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Input attempts to alter AI behavior or inject persistent instructions' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/data_poisoning.py
"""Detect data poisoning / model manipulation attempts."""

import time

from open_guardrail.core import GuardResult

_PATTERNS = [
    "training data", "fine-tune on", "fine tune on",
    "inject into model", "backdoor", "trigger phrase",
    "adversarial example", "poison the model",
    "corrupt the dataset", "manipulate training",
]


class _DataPoisoning:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "data-poisoning"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        matched = [
            p for p in _PATTERNS if p in lower
        ]
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="data-poisoning",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Data poisoning attempt detected"
                if triggered else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "matched": matched,
            } if triggered else None,
        )


def data_poisoning(
    *, action: str = "block",
) -> _DataPoisoning:
    return _DataPoisoning(action=action)
